### DISCUSSION TITLE
Why do I get an unclosed ssl-socket warning when committing/saving to a project?

### DISCUSSION THREAD

**Discussion by hendriklaue** [python-gitlab/python-gitlab#1553] [Q&A]
I'm commiting a file in a GitLab repository:
```
  gitlab.GitLab(URL, TOKEN).projects.get(ID)
  fileToBeUpdated = gitlabProject.files.get(
    file_path=f"{fileToBeUpdatedName}",
    ref=os.getenv('STORAGE_WORKING_BRANCH'))
  fileToBeUpdated.content = json.dumps(newData)
  fileToBeUpdated.save(branch=os.getenv('STORAGE_WORKING_BRANCH'), commit_message=f"Updated {fileToBeUpdatedName}")
```
When running my tests on this, I'll get a warning:
> /Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/http/cookiejar.py:1234: ResourceWarning: unclosed <ssl.SSLSocket fd=6, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('........', ...), raddr=('.....', 443)>
> yield from deepvalues(obj)
> ResourceWarning: Enable tracemalloc to get the object allocation traceback
> 

I resolved a similar issue by closing the GitLab-Connection using ```.sesssion.close()``` but this does not seem to apply for committing/writing files. I also did not find a corresponding method for files or branches. 

What am I missing?

**Comment by nejch** [comment]
@hendriklaue I don't really have enough context here to be able to tell what the issue might be, I don't think you shared the entire snippet.

If the test framework needs to be convinced that you close the session, which seems to be the case, you can use a context manager for that:

```python
with gitlab.Gitlab(host, token) as gl:
    gitlabProject = gl.projects.get(ID)
    fileToBeUpdated = gitlabProject.files.get(
        file_path=f"{fileToBeUpdatedName}",
        ref=os.getenv('STORAGE_WORKING_BRANCH'))
    fileToBeUpdated.content = json.dumps(newData)
    fileToBeUpdated.save(branch=os.getenv('STORAGE_WORKING_BRANCH'), commit_message=f"Updated {fileToBeUpdatedName}")

```

https://python-gitlab.readthedocs.io/en/stable/api-usage.html#context-manager

Of course that will likely be done in some other place if you have a larger test suite.
